### DISCUSSION TITLE
Piecewise Polynomial Tracking Question

### DISCUSSION THREAD

**Discussion by Iko-Git** [USC-ACTLab/crazyswarm#429] [Q&A]
Hello!
Could you please guide me or refer me to a clear source whereby I can learn how to execute and simulate 7th degree piecewise polynomials on multi CFs starting from a certain waypoint in the space for every CF. Providing an example would be highly appreciated 

It is important to mention that all my agent's trajectories have the same shape of [figure8.csv](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/scripts/figure8.csv) example

**Comment by whoenig** [comment]
I am not entirely sure what you are looking for. Perhaps slides 16 to 18 in https://drive.google.com/file/d/15favAyrLLpC_O6nrAp-eIbZijFUMLgwV/view helps.

**Reply by Iko-Git** [reply]
Thanks for your response!

Let me try to clarify more.

I have multi UAVs, each one of them has a unique trajectory presented as a .csv (similar to the one provided in the link [https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/scripts/figure8.csv](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/scripts/figure8.csv)).

What I am aiming to do is:
- Set a start waypoint for every agent, then let every agent hover to its start waypoint
- Feed the agent with its .csv trajectory
- See how the mission is performed in a simulator 

Any guidance or example providing would be highly appreciated.

**Reply by jpreiss** [reply]
How is this different from the example in `ros_ws/src/crazyswarm/scripts/figure8_csv.py` or `ros_ws/src/crazyswarm/scripts/swap6v.py`?

**Reply by Iko-Git** [reply]
Thanks for your response @jpreiss 
I am trying to run `python os_ws/src/crazyswarm/scripts/swap6v.py --sim` but it gives me the following message

```
Warning: No joystick found!
Traceback (most recent call last):
  File "swap6v.py", line 30, in <module>
    cfs = [allcfs.crazyfliesById[i] for i in ids]
KeyError: 15
```
Any idea how to treat it would be highly appreciated.

with respect to `figure8_csv`. right! it works, for single agent, but I am aiming to perform the described scenario on multi agents

**Reply by whoenig** [reply]
You need to update the ids in https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/scripts/swap6v.py#L27, according to the Crazyflies that are enabled in your crazyflies.yaml.

**Reply by Iko-Git** [reply]
I am trying to update it via nano or gedit but I am facing 
```
bash: gedit: command not found
bash: nano: command not found
```

**Reply by jpreiss** [reply]
Helping users install/configure text editors is outside the scope of Crazyswarm's support. If you believe this problem is somehow connected to the Crazyswarm package, then please explain why.

**Reply by Iko-Git** [reply]
I am not sure if the problem is somehow connected to the Crazyswarm package. However, I am able to use nano and gedit in the normal case. 
In the case of Crazyswarm package, what I am doing is the following:
`user: ~/crazyswarm_docker$ ./start.sh `
`root@user:/crazyswarm/ros_ws/src/crazyswarm/scripts# nano swap6v.py`
There I am facing 
`bash: nano: command not found`

**Reply by jpreiss** [reply]
It sounds like the problem is related to using Docker. Sorry I can't help beyond that, as I don't use Docker myself. I also don't know if Docker will be compatible with ROS and the radio hardware for running your experiment on real CFs.

**Reply by Iko-Git** [reply]
I could run the` swap6v.py` properly. The idea is that` swap6v.py` is performing an additional reverse navigation which I am aiming not to do.
Basically, what I am aiming to do is the multi agent following scenario:
- Hovering
- each agent navigates to certain waypoint in the space
- executing the trajectory from there
- waiting till all agents finish there trajectory
- landing (no coming back to the initial configurations)

I'd mention that in my case, the period of executing a segment of a piecewise for any agent in the mission is same for all agents (e.g. agent1 has 5 segments each one of them is performed in 2 seconds, consequently full mission 10 seconds. agent2 has 3 segments each one of them is performed again in 2 seconds, consequently its full mission takes 6 seconds)

**Reply by jpreiss** [reply]
That will be straightforward to implement. You can start with `swap6v.py` and make a few modifications for your scenario.

**Reply by Iko-Git** [reply]
Thanks for your comment. I am generating a trajectory on Matlab. The trajectory is generated as .csv file which has the same structure as the .csvs used in `swap6v.py` example. The problem is that the trajectory plot I am seeing on Matlab is different than the CF behavior in the simulator.
Below you can find an example on the csv I am using, and the plot of the trajectory.
![agent1_traj](https://user-images.githubusercontent.com/66323283/127742529-a1cd47a7-af29-432d-8303-9ab51220e99a.png)
[agent1.csv](https://github.com/USC-ACTLab/crazyswarm/files/6911299/agent1.csv)

**Reply by whoenig** [reply]
You can plot your csv file using https://github.com/whoenig/uav_trajectories/blob/master/scripts/plot_trajectory.py. Make sure that this looks like you expect, otherwise you might have used a slightly different encoding (order of polynomial coefficients?).

**Reply by Iko-Git** [reply]
Thanks for your remark. looks to be there was a mistake in the order of the coefficients!

Excuse me if the following two questions should not be posted under this discussion and please let me know if it has to be discussed in `uav_trajectories` repo.

1. In https://github.com/whoenig/uav_trajectories, trajectories can be generated based on a set of waypoints with constraints on the velocity and acceleration. E.g. `./genTrajectory -i examples/waypoints1.csv --v_max 1.0 --a_max 1.0 -o traj1.csv `. Is there any resource I can be referred to in order to study how the generated optimized trajectories consider the velocity and acceleration constraints?
2. Are the trajectories generated by the mentioned repo snap optimized?

Thanks again for your support!

**Reply by whoenig** [reply]
The repo is based on https://github.com/ethz-asl/mav_trajectory_generation, which in turn implements: C. Richter, A. Bry, and N. Roy, “Polynomial trajectory planning for aggressive quadrotor flight in dense indoor environments,” in International Journal of Robotics Research, Springer, 2016.
